engine/collection/synchronization/engine.go

```diff
@@ -361,9 +361,15 @@ func (e *Engine) pollHeight() {
 		return
 	}
 
+	nonce, err := rand.Uint64()
+	if err != nil {
+		e.log.Error().Err(err).Msg("nonce generation failed during pollHeight")
+		return
+	}
```

Comment:
We're extending an existing undesirable pattern in this engine, where errors are logged and swallowed. Could we:
- Document expected error returns in the `utils/rand` package
  - My understanding is, the `crypto/rand` might return an error if the system entropy pool runs low, though it's unlikely. We might want the node to continue operating in that case, though halting processes which require and fail to obtain randomness from the API. So, we should return a sentinel error for this case, and handle it here. 
- Add a TODO to this function godoc, or the log lines, that it should return an error rather than swallow+log errors

Reply:
Good point about documenting the expected error in `utils/rand`!

You are right about the swallow-and-log pattern here. Given it was commonly used in this file, I thought it was an intended pattern. 
Do you know if there is an issue to clean up this pattern under `/engine` so I can note the randomness errors ? I also noticed the other logged errors of the file miss a TODO statement.

> We might want the node to continue operating in that case

I don't think it's a good idea to continue operating. One way to attack some systems is to freeze/mess up with its entropy source. That can be fatal for some crypto protocols for instance. 
As I'm replacing `math/rand` by `crypto/rand`, there were a few cases where I swallowed the error and fell back to a simpler mechanism. The random jitters to balance the network load is one example, where it's unlikely to have a coordinated attack on multiple nodes to alter their system entropy. `crypto/rand` errors are also handled properly in many other places, so that a general failure will end up being properly caught and handled. 

Regarding in what cases `crypto/rand` outputs an error , it depends on the system where the code is operating. On a modern linux, my understanding of the package code is that it unlikely runs out of entropy, but if it does, the call would block till entropy is available again. The call may fail if there are other interrupting system calls, and in that case the node may have more blocking issues beyond randomness.

Reply:
> I don't think it's a good idea to continue operating.

I agree we shouldn't continue as if nothing was wrong -- what I meant initially was: don't crash, but terminate processes which fail to retrieve randomness. Though practically that's probably similar enough to crashing that we might as well just crash -- and it's much simpler. 

> there were a few cases where I swallowed the error and fell back to a simpler mechanism.

It sounds like the desired behaviour when a `utils/rand` call fails is:
- most of the time, we should consider this fatal, and throw an irrecoverable exception.
- in some cases, we might want to temporarily fallback to a simpler random source, however practically if we are failing to retrieve randomness system-wide another component which doesn't have a fallback will soon also hit an error and throw an irrecoverable exception.

Practically, I don't see much benefit to adding the fallbacks, if the node as a whole is likely to crash anyway. I feel like a policy of considering all errors from `crand` methods to be irrecoverable (not bothering with any fallbacks) is a good first pass at this.

Reply:
> Do you know if there is an issue to clean up this pattern under /engine so I can note the randomness errors ? I also noticed the other logged errors of the file miss a TODO statement.

I don't think there is one, made one for the sync engines here: https://github.com/onflow/flow-go/issues/4290